File: 1036-rotting-oranges/1036-rotting-oranges.py
from collections import deque

class Orange:
    def __init__(self, row, col, days):
        self.row = row
        self.col = col
        self.days = days

class Solution:
    def orangesRotting(self, grid):
        R, C = len(grid), len(grid[0])
        visited = [[False] * C for _ in range(R)]
        queue = deque()
        days, oranges = 0, 0

        for row in range(R):
            for col in range(C):
                if grid[row][col] != 0:
                    if grid[row][col] == 2:
                        queue.append(Orange(row, col, 0))
                        visited[row][col] = True
                    oranges += 1

        if oranges == 0:
            return 0  

        directions = [(0, -1), (0, 1), (-1, 0), (1, 0)]

        while queue:
            rotten = queue.popleft()
            oranges -= 1
            days = max(days, rotten.days)

            for dr, dc in directions:
                adjR, adjC = rotten.row + dr, rotten.col + dc

                if 0 <= adjR < R and 0 <= adjC < C:
                    if grid[adjR][adjC] == 1 and not visited[adjR][adjC]:
                        queue.append(Orange(adjR, adjC, rotten.days + 1))
                        visited[adjR][adjC] = True
                        grid[adjR][adjC] = 2  

        return -1 if oranges > 0 else days

File: 1036-rotting-oranges/1036-rotting-oranges.java
import java.util.*;

class Orange {
    int row, col, days;
    public Orange(int r, int c, int d) {
        this.row = r;
        this.col = c;
        this.days = d;
    }
}

class Solution {
    public int orangesRotting(int[][] grid) {
        int R = grid.length, C = grid[0].length;
        boolean[][] visited = new boolean[R][C];
        Queue<Orange> queue = new LinkedList<>();
        int days = 0, oranges = 0;

        for (int row = 0; row < R; row++) {
            for (int col = 0; col < C; col++) {
                if (grid[row][col] != 0) {
                    if (grid[row][col] == 2) {
                        queue.add(new Orange(row, col, 0));
                        visited[row][col] = true;
                    }
                    oranges++;
                }
            }
        }

        if (oranges == 0) return 0;  

        int[][] directions = {{0, -1}, {0, 1}, {-1, 0}, {1, 0}};

        while (!queue.isEmpty()) {
            Orange rotten = queue.poll();
            oranges--;
            days = Math.max(days, rotten.days);

            for (int index = 0; index < 4; index++) {
                int adjR = rotten.row + directions[index][0];
                int adjC = rotten.col + directions[index][1];

                if (adjR >= 0 && adjR < R && adjC >= 0 && adjC < C) {
                    if (grid[adjR][adjC] == 1 && !visited[adjR][adjC]) {
                        queue.add(new Orange(adjR, adjC, rotten.days + 1));
                        visited[adjR][adjC] = true;
                        grid[adjR][adjC] = 2;  
                    }
                }
            }
        }

        return oranges > 0 ? -1 : days;
    }
}